ext/liquid_c/context.c

```diff
@@ -41,25 +41,13 @@ VALUE context_find_variable(VALUE self, VALUE key, VALUE raise_on_not_found)
 {
     VALUE scope = Qnil, variable = Qnil;
 
-    VALUE scopes = rb_ivar_get(self, id_ivar_scopes);
-    Check_Type(scopes, T_ARRAY);
+    scope = rb_ivar_get(self, id_ivar_scope);
+    Check_Type(scope, T_HASH);
 
-    for (long i = 0; i < RARRAY_LEN(scopes); i++) {
-        VALUE this_scope = RARRAY_AREF(scopes, i);
-        if (RB_LIKELY(TYPE(this_scope) == T_HASH)) {
-            // Does not invoke any default value proc, this is equivalent in
-            // cost and semantics to #key? but loads the value as well
-            variable = rb_hash_lookup2(this_scope, key, Qundef);
-            if (variable != Qundef) {
-                scope = this_scope;
-                goto variable_found;
-            }
-        } else if (RTEST(rb_funcall(this_scope, id_has_key, 1, key))) {
-            // Slow path: It is valid to pass a non-hash value to Liquid as a
-            // scope if it supports #key? and #[]
-            variable = rb_funcall(this_scope, id_aref, 1, key);
-            goto variable_found;
-        }
+    variable = rb_hash_lookup2(scope, key, Qnil);
```

Comment:
To make this work identically to `#key?` as used here https://github.com/Shopify/liquid/pull/1090/files#diff-d97dbf3bc11085a7aa13ee71b9078b85R146 it should use `Qundef`. If the scope explicitly has the variable set as nil it should not be returned here.

Reply:
Yea good catch, I did this wrong on the Ruby PR as well I guess.

Reply:
Ugh my comment was backwards from what I was thinking... if the scope explicitly has the variable set as nil it _should_ be returned, and you have this right in the Ruby PR. That's how it worked before your change as well. The C PR is different though, it will not return the nil value.